runtime/src/storage/iouring.rs

```diff
@@ -0,0 +1,406 @@
+
+        Ok((
+            Blob::new(partition.into(), name, file, self.io_sender.clone()),
+            len,
+        ))
+    }
+
+    async fn remove(&self, partition: &str, name: Option<&[u8]>) -> Result<(), Error> {
+        let path = self.storage_directory.join(partition);
+        if let Some(name) = name {
+            let blob_path = path.join(hex(name));
+            fs::remove_file(blob_path)
+                .map_err(|_| Error::BlobMissing(partition.into(), hex(name)))?;
+        } else {
+            fs::remove_dir_all(path).map_err(|_| Error::PartitionMissing(partition.into()))?;
+        }
+        Ok(())
+    }
+
+    async fn scan(&self, partition: &str) -> Result<Vec<Vec<u8>>, Error> {
+        let path = self.storage_directory.join(partition);
+
+        let entries =
+            std::fs::read_dir(&path).map_err(|_| Error::PartitionMissing(partition.into()))?;
+
+        let mut blobs = Vec::new();
+        for entry in entries {
+            let entry = entry.map_err(|_| Error::ReadFailed)?;
+            let file_type = entry.file_type().map_err(|_| Error::ReadFailed)?;
+
+            if !file_type.is_file() {
+                return Err(Error::PartitionCorrupt(partition.into()));
+            }
+
+            if let Some(name) = entry.file_name().to_str() {
+                let name = from_hex(name).ok_or(Error::PartitionCorrupt(partition.into()))?;
+                blobs.push(name);
+            }
+        }
+
+        Ok(blobs)
+    }
+}
+
+pub struct Blob {
+    /// The partition this blob lives in
+    partition: String,
+    /// The name of the blob
+    name: Vec<u8>,
+    /// The underlying file
+    file: Arc<File>,
+    /// Where to send IO operations to be executed
+    io_sender: mpsc::Sender<(SqueueEntry, oneshot::Sender<i32>)>,
+}
+
+impl Clone for Blob {
+    fn clone(&self) -> Self {
+        Self {
+            partition: self.partition.clone(),
+            name: self.name.clone(),
+            file: self.file.clone(),
+            io_sender: self.io_sender.clone(),
+        }
+    }
+}
+
+impl Blob {
+    fn new(
+        partition: String,
+        name: &[u8],
+        file: File,
+        io_sender: mpsc::Sender<(SqueueEntry, oneshot::Sender<i32>)>,
+    ) -> Self {
+        Self {
+            partition,
+            name: name.to_vec(),
+            file: Arc::new(file),
+            io_sender,
+        }
+    }
+}
+
+impl crate::Blob for Blob {
+    async fn read_at(&self, buf: &mut [u8], offset: u64) -> Result<(), Error> {
+        let fd = types::Fd(self.file.as_raw_fd());
+        let mut total_read = 0;
+
+        let mut io_sender = self.io_sender.clone();
+        while total_read < buf.len() {
+            // Figure out how much is left to read and where to read into
+            let remaining = &mut buf[total_read..];
+            let offset = offset + total_read as u64;
+
+            // Create an operation to do the read
```

Comment:
If the future is cancelled before data is written, this could be a write-after-free (it isn't "cancel safe"): https://theincredibleholk.org/blog/2022/03/25/perspectives-on-async-cancellation

The only "cancel safe" solution AFAICT is for iouring to take ownership of the Vec?

While this may not be a problem for storage, we'll need to reckon with this to implement "timeouts": https://github.com/commonwarexyz/monorepo/blob/5abb303dee9e2ca7beeb1e033efc9f28226a833c/runtime/src/network/tokio.rs#L39

Reply:
WIll address in separate PR

Reply:
Resolved by https://github.com/commonwarexyz/monorepo/pull/933